### PR TITLE
Updates IonWriter method implementations to throw if value is undefined

### LINKED ISSUE #456: writeType(...) methods should not accept undefined
`Writer`'s `writeType` methods ([`writeTimestamp`](https://github.com/amzn/ion-js/blob/31871dc0e9404bef52037ae22c13d8aa49839362/src/IonBinaryWriter.ts#L236), [`writeString`](https://github.com/amzn/ion-js/blob/31871dc0e9404bef52037ae22c13d8aa49839362/src/IonBinaryWriter.ts#L214), etc) all accept `undefined` as a synonym for `null`. This means that calling `writeInt()` (with no parameters) produces `null.int`, which is rather surprising. Passing an `undefined` should probably cause an `Error` to be thrown.

### PR DESCRIPTION
**Pull request by pbcornell**
Resolves #456 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/IonBinaryWriter.ts [hunk 1]
```diff
@@ -96,6 +96,7 @@ export class BinaryWriter extends AbstractWriter {
     }
 
     writeBlob(value: Uint8Array): void {
+        _assertDefined(value);
         this.checkWriteValue();
         if (value === null || value === undefined) {
```
Comment (zslayton) [inline src/IonBinaryWriter.ts]: There are quite a few places where the new call to `_assertDefined(value)` happens before an 

```typescript
if (value === null || value === undefined) {
//...
}
```

The above can now be simplified to `if (value === null)`.
```

File: test/IonWriter.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    writeNull(type: IonType): void { }
+    writeString(value: string | null): void { }
+    writeSymbol(value: string | null): void { }
+    writeTimestamp(value: Timestamp | null): void { }
+    writeValue(reader: Reader): void { }
+    writeValues(reader: Reader): void { }
+}
+
+describe('IonWriter interface', () => {
+    let testMethods = Reflect.ownKeys(TestWriter.prototype)
```
Comment (zslayton) [inline test/IonWriter.ts]: Whoa, fancy!

File: test/IonWriter.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+ */
+
+import {assert} from "chai";
+import * as ion from "../src/IonTests";
+import {Decimal, IonType, IonTypes, Reader, Timestamp} from "../src/IonTests";
+import JSBI from "jsbi";
+
+// This dummy impl exists as a substitute for reflection directly against
+// the Writer interface (which TypeScript doesn't currently support)
+class TestWriter implements ion.Writer {
```
Comment (zslayton) [inline test/IonWriter.ts]: This file/class might be worth naming something more specific. `UndefinedValueWriter`?

File: src/IonBinaryWriter.ts (hunk 1 above)
Comment (pbcornell) [reply src/IonBinaryWriter.ts]: indeed, simplified

File: test/IonWriter.ts (hunk 3 above)
Comment (pbcornell) [reply test/IonWriter.ts]: Used generic terms with the idea that other tests against the `Writer` interface could go here, but I've renamed the file to `IonWriterUndefinedParameters`, and renamed `TestWriter` to `NoopWriter`.  Hope that helps a bit

**Pull request merged**